### PR TITLE
Allow `preserve_rng_state=True` when torch.compile + selective checkpointing + CUDA

### LINKED ISSUE #113717: `preserve_rng_state=True` in torch.utils.checkpoint causes error when used with torch.compile + selective checkpointing + CUDA
### 🐛 Describe the bug

As titled. Error log: https://gist.github.com/yf225/4a09fe4b12a482df81da94c5f22391b5
Repro by:
1. Changing input tensors in `test/dynamo/test_activation_checkpointing.py` test_compile_selective_checkpoint_gemm_only to CUDA.
2. Running `pytest -vs test/dynamo/test_activation_checkpointing.py::ActivationCheckpointingViaTagsTests::test_compile_selective_checkpoint_gemm_only`.

Root cause:
When `preserve_rng_state=True`, we let AOTAutograd trace through `torch.random.fork_rng` op, and the tracing doesn't work under CUDA. But since we are already doing RNG functionalization at Inductor level, we don't actually need to trace this `fork_rng` op.

### Versions

PyTorch nightly

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by yf225**
Fixes https://github.com/pytorch/pytorch/issues/113717.

When `preserve_rng_state=True`, we let AOTAutograd trace through `torch.random.fork_rng` op, and the tracing doesn't work under CUDA, hence the original error reported in the issue.

But since we are already doing RNG functionalization at Inductor level, we don't actually need to trace this `fork_rng` op. So we should just rewrite `preserve_rng_state` to False when we are using torch.compile (and let Inductor do its RNG functionalization which it's already been doing).

cc @soulitzer @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_higher_order_ops/wrap.py [hunk 1]
```diff
@@ -140,6 +141,11 @@ def __call__(self, gmod, *args, **kwargs):
             # And we ensure that AOT Autograd traces through the non reentrant
             # version of checkpointing.
             kwargs["use_reentrant"] = False
+            # preserve_rng_state is set to False because we want to prevent AOTAutograd from tracing through
+            # `torch.random.fork_rng` op (which is not supported yet under CUDA).
+            # This doesn't mean that we don't preserve RNG state. Instead, we will always preserve RNG state
+            # (i.e. do RNG functionalization via `replace_random_passes` in Inductor) regardless of this flag.
```
Comment (wanchaol) [inline torch/_higher_order_ops/wrap.py:147]: is this always true? I am looking at the https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/replace_random.py#L94, and it seems to me that it would replace the randn ops with inductor prim rand, but the seed are always newly created seed?

**Review by wanchaol — COMMENTED** [review]
Left a question, I wonder if we could add some tests to verify this approach, i.e. compile `preserve_rng_ops=True` behavior matches eager?

File: torch/_higher_order_ops/wrap.py (hunk 1 above)
Comment (yf225) [reply torch/_higher_order_ops/wrap.py:147]: the seed is newly created at compiled time, and it's then hardcoded into the graph. So at runtime we always reuse the same seed for the same random op.

**Comment by yf225** [comment]
> Left a question, I wonder if we could add some tests to verify this approach, i.e. compile `preserve_rng_ops=True` behavior matches eager?

I believe the best random op to test is `dropout`, but we need to be able to run in-place ops in the selective checkpointing region in order to have a unit test that uses dropout, which is blocked by https://github.com/pytorch/pytorch/issues/113737 :( I will figure that one out first before making progress on this.

File: torch/_higher_order_ops/wrap.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_higher_order_ops/wrap.py:147]: hmm isn't RNG functionalization controlled by this flag? Which looks like it's off by default - do we assert anywhere that it must be set when those code runs? https://github.com/pytorch/pytorch/blob/main/torch/_functorch/config.py#L16

Comment (yf225) [reply torch/_higher_order_ops/wrap.py:147]: It's complicated: AOTAutograd RNG functionalization is controlled by this flag that you mentioned (https://github.com/pytorch/pytorch/blob/b526aae95a7a9090ba7b628ce36d6e5d254a171e/torch/_functorch/config.py#L16), but Inductor RNG functionalization is done here https://github.com/pytorch/pytorch/blob/b526aae95a7a9090ba7b628ce36d6e5d254a171e/torch/_inductor/fx_passes/joint_graph.py#L277 and not controlled by that flag.

For both AC and SAC, we only use Inductor RNG functionalization, unless the functorch functionalize_rng_ops flag is on. (It is decided here: https://github.com/pytorch/pytorch/blob/b526aae95a7a9090ba7b628ce36d6e5d254a171e/torch/_dynamo/utils.py#L1976-L1980)

File: test/dynamo/test_activation_checkpointing.py [hunk 2]
```diff
@@ -642,37 +715,56 @@ def selective_checkpointing_context_fn():
...
+            # If in the future there is change to the partitioner that causes `dropout` to not be recomputed,
+            # the partitioner should expose an API for user to control what ops must be recomputed,
+            # so that we can ask the partitioner to always recompute this `dropout` op
+            # so that this unit test can still work.
+            freqs=[4, 0, 1],
+            ops=[
+                torch.ops.aten.mm.default,
+                torch.ops.aten.sigmoid.default,
+                torch.ops.aten.native_dropout.default,
+            ],
```
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:758]: Here we test that `dropout` is called during recompute.

File: test/dynamo/test_activation_checkpointing.py [hunk 3]
```diff
@@ -642,37 +715,56 @@ def selective_checkpointing_context_fn():
...
         )
         backend = aot_autograd(
             fw_compiler=fw_compiler,
             bw_compiler=bw_compiler,
             partition_fn=min_cut_rematerialization_partition,
         )
         self._validate(fn, backend, x, y)
+        if torch.cuda.is_available():
+            self._compare_orig_and_checkpointed_fns(gn, fn, x, y)
```
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:767]: Here we test that the original (non-checkpointed) version of the function gives the same output and gradients as the checkpointed version of the function.

File: test/dynamo/test_activation_checkpointing.py [hunk 4]
```diff
@@ -642,37 +715,60 @@ def selective_checkpointing_context_fn():
...
             return torch.utils.checkpoint.checkpoint(
                 gn,
-                torch.sin(x),
+                x,
                 y,
                 use_reentrant=False,
+                # Regardless of whether `preserve_rng_state` is True or False,
+                # we will always preserve RNG state when using `torch.compile`.
+                # Setting to True here just to show that True works.
+                preserve_rng_state=True,
```
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:733]: This shows that `preserve_rng_state=True` works now.

File: test/dynamo/test_activation_checkpointing.py [hunk 5]
```diff
@@ -625,54 +693,68 @@ def fn(x, y):
...
+                return torch.sigmoid(torch.dropout(torch.sigmoid(x), p=0.5, train=True))
+
+            def fn(x):
+                return torch.utils.checkpoint.checkpoint(
+                    gn,
+                    x,
+                    use_reentrant=False,
+                    # Regardless of whether `preserve_rng_state` is True or False,
+                    # we will always preserve RNG state when using `torch.compile`.
+                    preserve_rng_state=preserve_rng_state,
```
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:723]: This shows that `preserve_rng_state=True` works now.

File: test/dynamo/test_activation_checkpointing.py [hunk 6]
```diff
@@ -625,54 +693,69 @@ def fn(x, y):
...
-            fw_compiler=fw_compiler,
-            bw_compiler=bw_compiler,
-            partition_fn=min_cut_rematerialization_partition,
-        )
-        self._validate(fn, backend, x, y)
+            # NOTE: when `preserve_rng_state` is False, gradient will mismatch between torch.compile and eager,
+            # because eager version doesn't preserve RNG state while torch.compile still does.
+            # Hence when `preserve_rng_state` is False, we skip the output and gradient comparison
+            # between torch.compile and eager.
+            self._validate(fn, backend, x, skip_check=not preserve_rng_state)
```
Comment (yf225) [inline test/dynamo/test_activation_checkpointing.py:757]: When `preserve_rng_state=True`, we don't skip check, and we verify that eager and compiled output and gradients match.

File: test/dynamo/test_activation_checkpointing.py [hunk 7]
```diff
@@ -625,54 +693,69 @@ def fn(x, y):
...
-            ops=[torch.ops.aten.mm.default, torch.ops.aten.sigmoid.default],
-        )
-        backend = aot_autograd(
-            fw_compiler=fw_compiler,
-            bw_compiler=bw_compiler,
-            partition_fn=min_cut_rematerialization_partition,
-        )
-        self._validate(fn, backend, x, y)
+            # NOTE: when `preserve_rng_state` is False, gradient will mismatch between torch.compile and eager,
+            # because eager version doesn't preserve RNG state while torch.compile still does.
```
Comment (wanchaol) [inline test/dynamo/test_activation_checkpointing.py:754]: hmm I would expect if eager says "preserve_rng_state=False", then torch.compile should respect that?

**Review by wanchaol — APPROVED** [review]
LGTM, thanks for adding tests! have one question only

File: test/dynamo/test_activation_checkpointing.py (hunk 7 above)
Comment (yf225) [reply test/dynamo/test_activation_checkpointing.py:754]: yeah I am a little unsure what's the use case for `preserve_rng_state=False` in eager and if we should support that in compile. For now I just mention that under compile, `preserve_rng_state` is always True. I believe we can change this approach based on user's feedback
